mesonbuild/build.py

```diff
@@ -1969,6 +1969,11 @@ def __init__(
             environment: environment.Environment,
             compilers: T.Dict[str, 'Compiler'],
             kwargs):
+        # See https://github.com/rust-lang/rust/issues/110460
+        if 'rust' in compilers and kwargs.get('rust_crate_type', 'rlib') == 'rlib' and any(c in name for c in {'-', ' '}):
+            mlog.warning(f'Rust rlib static library target {name} contains spaces or dashes. '
+                         'This is not supported and these characters will be replaced by underscores')
+            name = name.replace('-', '_').replace(' ', '_')
```

Comment:
Can it be moved to BuildTarget base class rather than duplicating in SharedLibrary and StaticLibrary?

`compilers` here is all compilers from current project, not only those needed for this target. You can only know if this target uses rust by using `self.uses_rust()` after `process_compilers()` has been called, which is later from Interpreter.add_target().

Reply:
> Can it be moved to BuildTarget base class rather than duplicating in SharedLibrary and StaticLibrary?

Could be moved to the base class too, yes. But then the base class would have to check if it's a `SharedLibrary` or a `StaticLibrary` for the defaulting rules... which seems rather ugly.

> `compilers` here is all compilers from current project, not only those needed for this target. You can only know if this target uses rust by using `self.uses_rust()` after `process_compilers()` has been called, which is later from Interpreter.add_target().

Yeah see comment above. Where would I add this code then? `post_init` would work of course, but between `__init__` and `post_init` there could be a lot of code that already makes use of `name`, and if it doesn't do now it might do so in the future.

Reply:
Maybe we can do it in `mesonbuild/interpreter/interpreter.py:add_target()`? We already do some target name checking there. We could do this check after `tobj.process_compilers_late()` is called.

But seeing how error-prone this might get, and how we might be stuck with this behaviour forever, I am now leaning towards "let's just make it an error and explain that it's a rustc constraint".

Reply:
I.e. you'd want to go back to the initial version of the PR and extend the error message a bit?

Reply:
Yes

Reply:
Changed accordingly. Do you have suggestions how to improve the message?

Reply:
> But seeing how error-prone this might get, and how we might be stuck with this behaviour forever, I am now leaning towards "let's just make it an error and explain that it's a rustc constraint".

+1

> But then the base class would have to check if it's a SharedLibrary or a StaticLibrary for the defaulting rules... which seems rather ugly.

Not sure it matters which subclass it is: `if self.rust_crate_type in {'rlib', 'dylib', 'proc-macro} and any(c in self.name for c in ['-', ' ']):` is good in all cases, no?

But reading a bit that code, I think that would require a bit more of refactoring because self.rust_crate_type itself only has its final value in post_init() of the subclass... and now that I read even more code in build.py I see that self.compilers is used in many places before it gets initialized... this whole thing needs further refactoring sadly.

So LGTM for now :)

Reply:
Yes that was also my conclusion from going through this code :) AFAIU this will at least partially be part of the typing refactoring.